File: src/lib/augment-math-delimiters.ts
/**
 * 분석 JSON 등에서 모델이 `$` 없이 LaTeX를 섞어 줄 때 KaTeX가 동작하도록
 * `$…$` 를 삽입합니다.
 * 이미 `$`, `$$`, `\(`, `\[` 가 있으면 손대지 않습니다.
 */

/** `{` 시작 위치에서 짝 되는 `}` 직후 인덱스 */
function consumeBalancedBrace(s: string, openBrace: number): number {
  let depth = 1;
  let k = openBrace + 1;
  while (k < s.length && depth > 0) {
    const ch = s[k];
    if (ch === "{") depth++;
    else if (ch === "}") depth--;
    k++;
  }
  return k;
}

function skipSpaces(s: string, from: number): number {
  let k = from;
  while (k < s.length && /\s/.test(s[k]!)) k++;
  return k;
}

function tryParseFracAt(
  s: string,
  i: number,
): { full: string; end: number } | null {
  if (!s.startsWith("\\frac", i)) return null;
  let pos = skipSpaces(s, i + 5);
  if (s[pos] !== "{") return null;
  pos = consumeBalancedBrace(s, pos);
  pos = skipSpaces(s, pos);
  if (s[pos] !== "{") return null;
  pos = consumeBalancedBrace(s, pos);
  return { full: s.slice(i, pos), end: pos };
}

function tryParseSqrtAt(
  s: string,
  i: number,
): { full: string; end: number } | null {
  if (!s.startsWith("\\sqrt", i)) return null;
  let pos = skipSpaces(s, i + 5);
  if (s[pos] === "[") {
    const close = s.indexOf("]", pos + 1);
    if (close < 0) return null;
    pos = skipSpaces(s, close + 1);
  }
  if (s[pos] !== "{") return null;
  pos = consumeBalancedBrace(s, pos);
  return { full: s.slice(i, pos), end: pos };
}

/** `\frac`/`\sqrt`를 중첩 포함해 균형한 인자까지 한 덩어리로 감쌉니다 */
function wrapNestedConstructs(s: string): string {
  const stack: string[] = [];
  let pos = 0;
  while (pos < s.length) {
    if (s.startsWith("\\sqrt", pos)) {
      const p = tryParseSqrtAt(s, pos);
      if (p) {
        stack.push(p.full);
        pos = p.end;
        continue;
      }
    }
    if (s.startsWith("\\frac", pos)) {
      const p = tryParseFracAt(s, pos);
      if (p) {
        stack.push(p.full);
        pos = p.end;
        continue;
      }
    }
    pos++;
  }
  let t = s;
  for (const fragment of [...new Set(stack)].sort(
    (a, b) => b.length - a.length,
  )) {
    const wrapped = `$${fragment}$`;
    if (!fragment.includes("$")) {
      t = t.split(fragment).join(wrapped);
    }
  }
  return t;
}

/**
 * `\cos`, `\angle`, `\theta`, `QR=RS` 등 연속 LaTeX 친화 토큰을
 * 한 인라인 수식 블록으로 감쌉니다 (`$…$` 밖에서만).
 */
function consumeBareLatexRun(s: string, start: number): number {
  let j = start;
  while (j < s.length) {
    if (s[j] === "\\" && /[a-zA-Z]/i.test(s[j + 1] ?? "")) {
      j++;
      while (j < s.length && /[a-zA-Z]/i.test(s[j]!)) j++;
      if (s[j] === "*") j++;
      continue;
    }
    if (/\s/.test(s[j]!)) {
      const spStart = j;
      let sp = j;
      while (sp < s.length && /\s/.test(s[sp]!)) sp++;
      if (sp - spStart >= 2) break;
      if (sp < s.length && s[sp] === "\\") {
        j = sp;
        continue;
      }
      break;
    }
    const ch = s[j]!;
    // 한글·자모 시작이면 식 종료 ("…이므로 \angle BAC…"처럼 뒤에만 수식 있는 경우 처리)
    if (/[가-힣ㄱ-ㅎㅏ-ㅣ]/.test(ch)) break;
    if (/[0-9.]/.test(ch)) {
      while (j < s.length && /[0-9.]/.test(s[j]!)) j++;
      continue;
    }
    if (/[+\-=*/<>≠≤≥.(),;:!'|≈]/.test(ch)) {
      j++;
      continue;
    }
    if (/[A-Za-z]/.test(ch)) {
      while (j < s.length && /[A-Za-z]/.test(s[j]!)) j++;
      continue;
    }
    if (ch === "_" || ch === "^") {
      j++;
      if (s[j] === "{") {
        j = consumeBalancedBrace(s, j);
        continue;
      }
      if (/[0-9A-Za-z(]/.test(s[j] ?? "")) j++;
      continue;
    }
    if (ch === "{") {
      j = consumeBalancedBrace(s, j);
      continue;
    }
    break;
  }
  return j;
}

function wrapBackslashSequencesOutsideDelimiters(fragment: string): string {
  if (!fragment.includes("\\")) return fragment;
  let out = "";
  let i = 0;
  while (i < fragment.length) {
    if (fragment[i] === "\\" && /[a-zA-Z]/i.test(fragment[i + 1] ?? "")) {
      const end = consumeBareLatexRun(fragment, i);
      if (end > i) {
        const piece = fragment.slice(i, end).trimEnd();
        if (piece.length > 0 && /\\/.test(piece)) {
          out += `$${piece}$`;
          i = end;
          continue;
        }
      }
    }
    out += fragment[i]!;
    i++;
  }
  return out;
}

function transformOutsideExistingMath(
  s: string,
  fn: (frag: string) => string,
): string {
  if (!s.includes("$")) return fn(s);
  const chunks = s.split("$");
  return chunks
    .map((chunk, idx) => (idx % 2 === 0 ? fn(chunk) : chunk))
    .join("$");
}

function segmentHasBareLatex(s: string): boolean {
  if (!s.trim()) return false;
  return (
    /\^[0-9]|_[0-9a-z]|\)\\^|[²³⁴]/.test(s) ||
    /\\(?:[a-zA-Z]+\*?)/i.test(s)
  );
}

function augmentPlainConstructs(input: string): string {
  let t = input;
  t = t.replace(/\([^()]*\)\^[0-9]+/g, (m) => `$${m}$`);
  t = t.replace(
    /\b([a-zA-Z])\s*=\s*([a-zA-Z0-9+*/().\s-]+)/g,
    (full, v: string, rhs: string) => {
      const r = String(rhs).trim();
      if (!/[\^_]|\\|\\frac|\\sqrt/.test(r)) return full;
      return `$${v}=${r}$`;
    },
  );
  t = t.replace(/\b[a-zA-Z][a-zA-Z0-9']*\^[0-9]+\b/g, (m) => {
    if (m.includes("$")) return m;
    return `$${m}$`;
  });
  t = t.replace(/\b[a-zA-Z][a-zA-Z0-9']*_[0-9a-z]+\b/gi, (m) => {
    if (m.includes("$")) return m;
    return `$${m}$`;
  });
  t = t.replace(/([a-zA-Z])[²³⁴]/g, (fullMatch, letter: string) => {
    const sup = fullMatch.slice(-1);
    const map: Record<string, string> = { "²": "^2", "³": "^3", "⁴": "^4" };
    return `$${letter}${map[sup] ?? "^2"}$`;
  });
  t = t.replace(/\([^()]+\)[²³⁴]/g, (full) => {
    const sup = full.slice(-1);
    const map: Record<string, string> = { "²": "^2", "³": "^3", "⁴": "^4" };
    const base = full.slice(0, -1);
    return `$${base}${map[sup] ?? "^2"}$`;
  });
  return t;
}

export function augmentMathDelimiters(input: string): string {
  const s = input ?? "";
  if (!s.trim()) return s;

  const hasDelimiter = /\$|\\\(|\\\[/.test(s);
  if (hasDelimiter) return s;

  if (!segmentHasBareLatex(s)) return s;

  let t = transformOutsideExistingMath(s, augmentPlainConstructs);
  t = transformOutsideExistingMath(t, wrapNestedConstructs);
  t = transformOutsideExistingMath(t, wrapBackslashSequencesOutsideDelimiters);
  return t;
}

File: flutter_app/lib/utils/augment_math_delimiters.dart
/// 웹 `augment-math-delimiters.ts` 와 동일 — `$` 없이 섞인 LaTeX를 KaTeX/Math.tex 가 읽을 수 있게 감쌉니다.
int _consumeBalancedBrace(String s, int openBrace) {
  var depth = 1;
  var k = openBrace + 1;
  while (k < s.length && depth > 0) {
    final ch = s[k];
    if (ch == '{') {
      depth++;
    } else if (ch == '}') {
      depth--;
    }
    k++;
  }
  return k;
}

int _skipSpaces(String s, int from) {
  var k = from;
  while (k < s.length && RegExp(r'\s').hasMatch(s[k])) {
    k++;
  }
  return k;
}

({String full, int end})? _tryParseFracAt(String s, int i) {
  if (!s.startsWith(r'\frac', i)) return null;
  var pos = _skipSpaces(s, i + 5);
  if (pos >= s.length || s[pos] != '{') return null;
  pos = _consumeBalancedBrace(s, pos);
  pos = _skipSpaces(s, pos);
  if (pos >= s.length || s[pos] != '{') return null;
  pos = _consumeBalancedBrace(s, pos);
  return (full: s.substring(i, pos), end: pos);
}

({String full, int end})? _tryParseSqrtAt(String s, int i) {
  if (!s.startsWith(r'\sqrt', i)) return null;
  var pos = _skipSpaces(s, i + 5);
  if (pos < s.length && s[pos] == '[') {
    final close = s.indexOf(']', pos + 1);
    if (close < 0) return null;
    pos = _skipSpaces(s, close + 1);
  }
  if (pos >= s.length || s[pos] != '{') return null;
  pos = _consumeBalancedBrace(s, pos);
  return (full: s.substring(i, pos), end: pos);
}

String _wrapNestedConstructs(String s) {
  final stack = <String>[];
  var pos = 0;
  while (pos < s.length) {
    if (s.startsWith(r'\sqrt', pos)) {
      final parsed = _tryParseSqrtAt(s, pos);
      if (parsed != null) {
        stack.add(parsed.full);
        pos = parsed.end;
        continue;
      }
    }
    if (s.startsWith(r'\frac', pos)) {
      final parsed = _tryParseFracAt(s, pos);
      if (parsed != null) {
        stack.add(parsed.full);
        pos = parsed.end;
        continue;
      }
    }
    pos++;
  }

  var t = s;
  final unique = stack.toSet().toList()
    ..sort((a, b) => b.length.compareTo(a.length));
  for (final fragment in unique) {
    if (fragment.contains(r'$')) continue;
    t = t.split(fragment).join('\$$fragment\$');
  }
  return t;
}

int _consumeBareLatexRun(String s, int start) {
  var j = start;
  while (j < s.length) {
    if (s[j] == r'\' && j + 1 < s.length && RegExp(r'[a-zA-Z]').hasMatch(s[j + 1])) {
      j++;
      while (j < s.length && RegExp(r'[a-zA-Z]').hasMatch(s[j])) {
        j++;
      }
      if (j < s.length && s[j] == '*') j++;
      continue;
    }
    if (RegExp(r'\s').hasMatch(s[j])) {
      final spStart = j;
      var sp = j;
      while (sp < s.length && RegExp(r'\s').hasMatch(s[sp])) {
        sp++;
      }
      if (sp - spStart >= 2) break;
      if (sp < s.length && s[sp] == r'\') {
        j = sp;
        continue;
      }
      break;
    }
    final ch = s[j];
    if (RegExp(r'[가-힣ㄱ-ㅎㅏ-ㅣ]').hasMatch(ch)) break;
    if (RegExp(r'[0-9.]').hasMatch(ch)) {
      while (j < s.length && RegExp(r'[0-9.]').hasMatch(s[j])) {
        j++;
      }
      continue;
    }
    if (RegExp(r'''[+\-=*/<>≠≤≥.(),;:!'|≈]''').hasMatch(ch)) {
      j++;
      continue;
    }
    if (RegExp(r'[A-Za-z]').hasMatch(ch)) {
      while (j < s.length && RegExp(r'[A-Za-z]').hasMatch(s[j])) {
        j++;
      }
      continue;
    }
    if (ch == '_' || ch == '^') {
      j++;
      if (j < s.length && s[j] == '{') {
        j = _consumeBalancedBrace(s, j);
        continue;
      }
      if (j < s.length && RegExp(r'[0-9A-Za-z(]').hasMatch(s[j])) {
        j++;
      }
      continue;
    }
    if (ch == '{') {
      j = _consumeBalancedBrace(s, j);
      continue;
    }
    break;
  }
  return j;
}

String _wrapBackslashSequencesOutsideDelimiters(String fragment) {
  if (!fragment.contains(r'\')) return fragment;
  final buffer = StringBuffer();
  var i = 0;
  while (i < fragment.length) {
    if (fragment[i] == r'\' &&
        i + 1 < fragment.length &&
        RegExp(r'[a-zA-Z]').hasMatch(fragment[i + 1])) {
      final end = _consumeBareLatexRun(fragment, i);
      if (end > i) {
        final piece = fragment.substring(i, end).trimRight();
        if (piece.isNotEmpty && piece.contains(r'\')) {
          buffer.write('\$$piece\$');
          i = end;
          continue;
        }
      }
    }
    buffer.write(fragment[i]);
    i++;
  }
  return buffer.toString();
}

String _transformOutsideExistingMath(String s, String Function(String) fn) {
  if (!s.contains(r'$')) return fn(s);
  final chunks = s.split(r'$');
  final out = StringBuffer();
  for (var idx = 0; idx < chunks.length; idx++) {
    out.write(idx.isEven ? fn(chunks[idx]) : chunks[idx]);
    if (idx < chunks.length - 1) out.write(r'$');
  }
  return out.toString();
}

bool _segmentHasBareLatex(String s) {
  if (s.trim().isEmpty) return false;
  return RegExp(r'\^[0-9]|_[0-9a-z]|\)\^|[²³⁴]').hasMatch(s) ||
      RegExp(r'\\(?:[a-zA-Z]+\*?)', caseSensitive: false).hasMatch(s);
}

String _augmentPlainConstructs(String input) {
  var t = input;
  t = t.replaceAllMapped(RegExp(r'\([^()]*\)\^[0-9]+'), (m) => '\$${m[0]}\$');
  t = t.replaceAllMapped(
    RegExp(r'\b([a-zA-Z])\s*=\s*([a-zA-Z0-9+*/().\s-]+)'),
    (m) {
      final rhs = m.group(2)!.trim();
      if (!RegExp(r'[\^_]|\\|\\frac|\\sqrt').hasMatch(rhs)) {
        return m[0]!;
      }
      return '\$${m.group(1)}=$rhs\$';
    },
  );
  t = t.replaceAllMapped(RegExp(r'\b[a-zA-Z][a-zA-Z0-9\x27]*\^[0-9]+\b'), (m) {
    if (m[0]!.contains(r'$')) return m[0]!;
    return '\$${m[0]}\$';
  });
  t = t.replaceAllMapped(
    RegExp(r'\b[a-zA-Z][a-zA-Z0-9\x27]*_[0-9a-z]+\b', caseSensitive: false),
    (m) {
      if (m[0]!.contains(r'$')) return m[0]!;
      return '\$${m[0]}\$';
    },
  );
  t = t.replaceAllMapped(RegExp(r'([a-zA-Z])[²³⁴]'), (m) {
    const map = {'²': '^2', '³': '^3', '⁴': '^4'};
    final sup = m[0]!.substring(m[0]!.length - 1);
    return '\$${m.group(1)}${map[sup] ?? '^2'}\$';
  });
  t = t.replaceAllMapped(RegExp(r'\([^()]+\)[²³⁴]'), (m) {
    const map = {'²': '^2', '³': '^3', '⁴': '^4'};
    final sup = m[0]!.substring(m[0]!.length - 1);
    final base = m[0]!.substring(0, m[0]!.length - 1);
    return '\$${base}${map[sup] ?? '^2'}\$';
  });
  t = t.replaceAllMapped(
    RegExp(r'[\d.]+\s*\\(?:times|cdot)\s*[\d.]+', caseSensitive: false),
    (m) => '\$${m[0]}\$',
  );
  return t;
}

String augmentMathDelimiters(String input) {
  final s = input;
  if (s.trim().isEmpty) return s;

  final hasDelimiter = RegExp(r'\$|\\\(|\\\[').hasMatch(s);
  if (hasDelimiter) return s;

  if (!_segmentHasBareLatex(s)) return s;

  var t = _transformOutsideExistingMath(s, _augmentPlainConstructs);
  t = _transformOutsideExistingMath(t, _wrapNestedConstructs);
  t = _transformOutsideExistingMath(t, _wrapBackslashSequencesOutsideDelimiters);
  return t;
}
